core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java

```diff
@@ -0,0 +1,152 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+
+/**
+ * A skeletal {@link EndpointSelector} implementation. This abstract class implements the
+ * {@link #select(ClientRequestContext, ScheduledExecutorService, long)} method by listening to
+ * the change events emitted by {@link EndpointGroup} specified at construction time.
+ */
+public abstract class AbstractEndpointSelector implements EndpointSelector {
+
+    private final EndpointGroup endpointGroup;
+
+    /**
+     * Creates a new instance that selects an {@link Endpoint} from the specified {@link EndpointGroup}.
+     */
+    protected AbstractEndpointSelector(EndpointGroup endpointGroup) {
+        this.endpointGroup = requireNonNull(endpointGroup, "endpointGroup");
+    }
+
+    /**
+     * Returns the {@link EndpointGroup} being selected by this {@link EndpointSelector}.
+     */
+    protected final EndpointGroup group() {
+        return endpointGroup;
+    }
+
+    @Override
+    public final CompletableFuture<Endpoint> select(ClientRequestContext ctx,
+                                                    ScheduledExecutorService executor,
+                                                    long timeoutMillis) {
+        Endpoint endpoint = selectNow(ctx);
+        if (endpoint != null) {
+            return UnmodifiableFuture.completedFuture(endpoint);
+        }
+
+        final ListeningFuture listeningFuture = new ListeningFuture(ctx, executor);
+        endpointGroup.addListener(listeningFuture);
+
+        // Try to select again because the EndpointGroup might have been updated
+        // between selectNow() and addListener() above.
+        endpoint = selectNow(ctx);
+        if (endpoint != null) {
+            endpointGroup.removeListener(listeningFuture);
+            return UnmodifiableFuture.completedFuture(endpoint);
+        }
+
+        // Schedule the timeout task.
+        final ScheduledFuture<?> timeoutFuture =
+                executor.schedule(() -> listeningFuture.complete(null),
+                                  timeoutMillis, TimeUnit.MILLISECONDS);
+        listeningFuture.timeoutFuture = timeoutFuture;
+
+        // Cancel the timeout task if necessary, just in case listeningFuture is
+        // completed even before we assign timeoutFuture to listeningFuture.timeoutFuture.
```

Comment:
If so `listeningFuture.timeoutFuture = timeoutFuture;` this line should move down.
However, I'm wondering if there's a chance that `listeningFuture.isDone()` is `true` after we call `endpoint = selectNow(ctx);` when `endpoint` is `null`. 🤔

Reply:
Updated the comment to explain better.